2002-01-03  Gwenole Beauchesne  <gbeauchesne@mandrakesoft.com>

	* gcc/testsuite/gcc.c-torture/compile/20020103-1.c: New test.

2001-12-01  Olivier Hainque  <hainque@act-europe.fr>

	* unroll.c (loop_iterations): Give up on jumps with null
	JUMP_LABEL while scanning for multiple back edges.

--- gcc-2.95.2/gcc/testsuite/gcc.c-torture/compile/20020103-1.c.loop-iterations2	Thu Jan  3 18:17:02 2002
+++ gcc-2.95.2/gcc/testsuite/gcc.c-torture/compile/20020103-1.c	Thu Jan  3 18:16:46 2002
@@ -0,0 +1,32 @@
+typedef union instr
+{
+  void*   p;
+  int     i;
+} Instr;
+
+#define NEXT(step)		{__ip += (step); goto *(__ip->p);}
+#define BRA(offset)		{__ip += (offset); NEXT (0);}
+#define GET_ICODE_I32(offset)	(__ip[(offset) + 1].i)
+
+static void engine (void)
+{
+  Instr *__ip;
+  
+ lookupswitch:
+  {
+    int npairs = GET_ICODE_I32 (1);
+    unsigned i;
+    for (i = 0; i < npairs; i++)
+      {
+        int match = GET_ICODE_I32 (i*2 + 2);
+        if (match == 0)
+          {
+            int offset = GET_ICODE_I32 (i*2 + 3);
+            BRA(offset);
+          }  
+        if (match > 0)
+          break;
+      }
+  }
+  NEXT (0);
+}
--- gcc-2.95.2/gcc/unroll.c.loop-iterations2	Thu Jan  3 17:49:58 2002
+++ gcc-2.95.2/gcc/unroll.c	Thu Jan  3 18:08:12 2002
@@ -3706,19 +3706,32 @@
 
       do
 	{
-	  if (GET_CODE (temp) == JUMP_INSN
-	      /* Previous unrolling may have generated new insns not covered
-		 by the uid_luid array.  */
-	      && INSN_UID (JUMP_LABEL (temp)) < max_uid_for_loop
-	      /* Check if we jump back into the loop body.  */
-	      && INSN_LUID (JUMP_LABEL (temp)) > INSN_LUID (loop->top)
-	      && INSN_LUID (JUMP_LABEL (temp)) < INSN_LUID (loop->cont))
+	  if (GET_CODE (temp) == JUMP_INSN)
 	    {
-	      if (loop_dump_stream)
-		fprintf (loop_dump_stream,
-			 "Loop iterations: Loop has multiple back edges.\n");
-	      return 0;
+	      /* There are some kinds of jumps we can't deal with easily.  */
+	      if (JUMP_LABEL (temp) == 0)
+		{
+		  if (loop_dump_stream)
+		    fprintf
+		      (loop_dump_stream,
+		       "Loop iterations: Jump insn has null JUMP_LABEL.\n");
+		  return 0;
+		}
+
+	      if (/* Previous unrolling may have generated new insns not
+		     covered by the uid_luid array.  */
+		  INSN_UID (JUMP_LABEL (temp)) < max_uid_for_loop
+		  /* Check if we jump back into the loop body.  */
+		  && INSN_LUID (JUMP_LABEL (temp)) > INSN_LUID (loop->top)
+		  && INSN_LUID (JUMP_LABEL (temp)) < INSN_LUID (loop->cont))
+		{
+		  if (loop_dump_stream)
+		    fprintf 
+		      (loop_dump_stream,
+		       "Loop iterations: Loop has multiple back edges.\n");
+		  return 0;
+		}
 	    }
 	}
       while ((temp = PREV_INSN (temp)) != loop->cont);
